tests/test_transf_comp.py

```diff
@@ -1,29 +1,33 @@
-import unittest
+import pytest
 import numpy as np
 from cpmpy import *
 from cpmpy.transformations.flatten_model import flatten_constraint
 from cpmpy.transformations.comparison import only_numexpr_equality
 from cpmpy.expressions.variables import _IntVarImpl, _BoolVarImpl # to reset counters
 
-class TestTransfComp(unittest.TestCase):
-    def setUp(self):
-        _IntVarImpl.counter = 0
-        _BoolVarImpl.counter = 0
+def _generate_inputs():
+    ivs = intvar(1,9, shape=3, name="x")
+    for case in [
+            (min(ivs) == 3, "[min(x[0],x[1],x[2]) == 3]"),
+            (min(ivs) > 3, "[(min(x[0],x[1],x[2])) == (IV0), IV0 > 3]"),
+            (min(ivs) <= 3, "[(min(x[0],x[1],x[2])) == (IV0), IV0 <= 3]"),
+            (3 != max(ivs), "[(max(x[0],x[1],x[2])) == (IV0), IV0 != 3]"),
+            (3 > max(ivs), "[(max(x[0],x[1],x[2])) == (IV0), IV0 < 3]"),
+            (3 <= max(ivs), "[(max(x[0],x[1],x[2])) == (IV0), IV0 >= 3]"),
+            ]:
+        yield case
 
-    def test_only_numexpr_eq(self):
-        ivs = intvar(1,9, shape=3, name="ivs")
-        
-        cases = [(min(ivs) == 3, "[min(ivs[0],ivs[1],ivs[2]) == 3]"),
-                 (min(ivs) > 3, "[(min(ivs[0],ivs[1],ivs[2])) == (IV0), IV0 > 3]"),
-                 (min(ivs) <= 3, "[(min(ivs[0],ivs[1],ivs[2])) == (IV2), IV2 <= 3]"),
-                 (3 != max(ivs), "[(max(ivs[0],ivs[1],ivs[2])) == (IV4), IV4 != 3]"),
-                 (3 > max(ivs), "[(max(ivs[0],ivs[1],ivs[2])) == (IV6), IV6 < 3]"),
-                 (3 <= max(ivs), "[(max(ivs[0],ivs[1],ivs[2])) == (IV8), IV8 >= 3]"),
-                ]
+@pytest.fixture()
+def setup():
+    _IntVarImpl.counter = 0
+    _BoolVarImpl.counter = 0
+    yield
 
-        # test transformation
-        for (expr, strexpr) in cases:
-            self.assertEqual( str(only_numexpr_equality(flatten_constraint(expr))), strexpr )
-            self.assertTrue(Model(expr).solve())
+class TestTransform:
+    @pytest.mark.parametrize(("constraint", "expected"),_generate_inputs(), ids=str)
```

Comment:
looks like black magic to me... has some hidden assumption that the 2nd argument returns tuples of size 2?

and there is some even more magic 'ids=str' argument?

Reply:
Absolutely agree, this is all pytest magic (same as in test_constraints). The tuple is jsut because a test case is defined by those two arguments (as in prev. comment, I'd add the transform function if there were different ones tested!). The ids=str just names the unit test case (eg `test_transforms[min(x[0],x[1],x[2]) == 3-[min(x[0],x[1],x[2]) == 3]]`)

Reply:
Actually this `str=ids` is the default, so I've just removed it.